### ISSUE TITLE
You cannot call a method on a null-valued expression

### ISSUE THREAD

**Issue by EloiseJory** [AsBuiltReport/AsBuiltReport.VMware.vSphere#70]
**Describe the bug**
You cannot call a method on a null-valued expression  error after getting to this section:
VERBOSE: [ 16:24:39:335 ] [ Document ] - Processing section 'DSC - Trusted Colours' completed.
VERBOSE: [ 16:24:39:350 ] [ Document ] - Processing section 'Datastore Clusters' completed.
VERBOSE: [ 16:24:39:350 ] [ Document ] - Processing section 'Virtual Machines' started.
VERBOSE: [ 16:24:39:366 ] [ Document ] - Processing paragraph 'The following sections detail the co[..]'.
VERBOSE: [ 16:24:39:366 ] [ Document ] - Processing blank line.
New-AsBuiltReport : You cannot call a method on a null-valued expression.
At line:1 char:1
+ New-AsBuiltReport -Report VMware.vSphere -Target 172.22.102.150 -User ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,New-AsBuiltReport

**To Reproduce**
New-AsBuiltReport -Report VMware.vSphere -Target 172.22.xxx.yyy -Username username -Password password OutputPath 'd:\Reports\'

The licencing section is set to false in the json due to causing the report to fail.

**Expected behavior**
Report to be generated

**Screenshots**
VERBOSE: [ 16:24:39:335 ] [ Document ] - Processing section 'DSC - Trusted Colours' completed.
VERBOSE: [ 16:24:39:350 ] [ Document ] - Processing section 'Datastore Clusters' completed.
VERBOSE: [ 16:24:39:350 ] [ Document ] - Processing section 'Virtual Machines' started.
VERBOSE: [ 16:24:39:366 ] [ Document ] - Processing paragraph 'The following sections detail the co[..]'.
VERBOSE: [ 16:24:39:366 ] [ Document ] - Processing blank line.
New-AsBuiltReport : You cannot call a method on a null-valued expression.
At line:1 char:1
+ New-AsBuiltReport -Report VMware.vSphere -Target 172.22.xxx.yyy -User ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,New-AsBuiltReport

PS H:\> $global:error[1]
You cannot call a method on a null-valued expression.
At C:\Program
Files\WindowsPowerShell\Modules\AsBuiltReport.VMware.vSphere\Src\Public\Invoke-AsBuiltReport.VMware.vSphere.ps1:3421
char:64
+ ...             'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

PS H:\>

**System (please provide the following information about the system from which you are trying to generate a report):**
 - OS: server 2016
 - Windows PowerShell version [Provide output from the following command: `$PSVersionTable.PSVersion`]

PS H:\> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      14393  2969

AsBuiltReport Core 1.0.0 AsBuiltReport.Core

AsBuiltReport Report module name and version 1.1.0 AsBuiltReport.VMware.vSphere0.0]

**Comment by tpcarman** [comment]
@EloiseJory you seem to be using the vSphere report module from my [dev repo](https://github.com/tpcarman/AsBuiltReport.VMware.vSphere).

Have you tried using the module from [PSGallery](https://www.powershellgallery.com/packages/AsBuiltReport.VMware.vSphere/1.0.7)?

**Comment by EloiseJory** [comment]
Yes i initially tried the prod version and had these errors so thought I'd give the dev version a whirl..

**Comment by tpcarman** [comment]
I would suggest that you set all InfoLevels to 0 and then gradually increase the InfoLevel, one section at a time to help isolate where you are having the issue.

**Comment by EloiseJory** [comment]
Running the report withInfolevel 0 on all settings, lets the script complete, but it doesn’t output anything (As expected).

If Iset vCentre to 1, or 2, the script completes, and yields an empty report.

Setting InfoLevel to 3 for vCenter, as below, the report crashes, with the same error as before.

**Comment by zpatrickson** [comment]
I'm facing the exact same issue. Have you figured out a solution to this issue?

**Comment by EloiseJory** [comment]
No resolution yet @zpatrickson   Was ok on vsphere 6.5 but won't work on 6.7 u2

**Comment by tpcarman** [comment]
Hi @zpatrickson are you able to provide some more information regarding your vSphere environment and the AsBuiltReport module versions? Do you have anything unusual or unique in your environment?

Would you be open to doing a webex (or similar) session to allow me to investigate further?

**Comment by zpatrickson** [comment]
Hi @tpcarman, sure we can do a webex. I will send you a PM with my work email address. I'm using version 1.0.7 . Nothing unusual in my environment. My ESXi is 5.5 and planning to do an upgrade. I just want to document it before hand.

**Comment by tpcarman** [comment]
@zpatrickson I am in UTC +10 for the next 1.5 weeks. Late evening my time, early morning your time probably works best for me. We can either do webex or do LiveShare within VS Code. Let me know if this works for you.

**Comment by tpcarman** [comment]
@zpatrickson Send me your email address via the [contact page](https://www.asbuiltreport.com/support/contact/).

**Comment by tpcarman** [comment]
@zpatrickson @EloiseJory hi guys, would you mind connecting to your vCenter with PowerCLI and running the following;
```powershell
$VMs = Get-VM
$VMInfo = foreach ($VM in $VMs) {
	$VMView = $VM | Get-View
	[PSCustomObject]@{
		'Name' = $VM.Name
		'Power State' = Switch ($VM.PowerState) {
			'PoweredOn' { 'On' }
			'PoweredOff' { 'Off' }
			default { $VM.PowerState }
		}
		'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
			$null { '--' }
			default { $VMView.Summary.Guest.IpAddress }
		}
		'vCPUs' = $VM.NumCpu
		'Memory GB' = [math]::Round(($VM.MemoryGB), 0)
		'Provisioned GB' = [math]::Round(($VM.ProvisionedSpaceGB), 0)
		'Used GB' = [math]::Round(($VM.UsedSpaceGB), 0)
		'HW Version' = ($VM.HardwareVersion).Replace('vmx-', 'v')
		'VM Tools Status' = Switch ($VMView.Summary.Guest.ToolsStatus) {
			'toolsOld' { 'Old' }
			'toolsOK' { 'OK' }
			'toolsNotRunning' { 'Not Running' }
			'toolsNotInstalled' { 'Not Installed' }
			default { $VMView.Guest.ToolsStatus }
		}         
	}
}
Write-Output $VMInfo | ft
```
Please let me know how this goes.

**Comment by EloiseJory** [comment]
> @zpatrickson @EloiseJory hi guys, would you mind connecting to your vCenter with PowerCLI and running the following;

I should be able to test it tomorrow - will let you know !!

thanks

Eloise.

**Comment by EloiseJory** [comment]
Hi
All I get is errors:


You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull

**Comment by tpcarman** [comment]
What about this?
```
$VMs = Get-VM
$VMInfo = foreach ($VM in $VMs) {
	$VMView = $VM | Get-View
	[PSCustomObject]@{
		'Name' = $VM.Name
		'Power State' = Switch ($VM.PowerState) {
			'PoweredOn' { 'On' }
			'PoweredOff' { 'Off' }
			default { $VM.PowerState }
		}
		'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
			$null { '--' }
                        '' { '--' }
			default { $VMView.Summary.Guest.IpAddress }
		}
		'vCPUs' = $VM.NumCpu
		'Memory GB' = [math]::Round(($VM.MemoryGB), 0)
		'Provisioned GB' = [math]::Round(($VM.ProvisionedSpaceGB), 0)
		'Used GB' = [math]::Round(($VM.UsedSpaceGB), 0)
		'HW Version' = ($VM.HardwareVersion).Replace('vmx-', 'v')
		'VM Tools Status' = Switch ($VMView.Summary.Guest.ToolsStatus) {
			'toolsOld' { 'Old' }
			'toolsOK' { 'OK' }
			'toolsNotRunning' { 'Not Running' }
			'toolsNotInstalled' { 'Not Installed' }
			default { $VMView.Guest.ToolsStatus }
		}         
	}
}
Write-Output $VMInfo | ft
```
If that still doesn't work, try removing these lines and trying again
```
'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
			$null { '--' }
                        '' { '--' }
			default { $VMView.Summary.Guest.IpAddress }
		}
```

**Comment by EloiseJory** [comment]
> What about this?
First attempt was a heap of these again


`You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:11 char:26
+         'IP Address' = Switch ($VMView.Summary.Guest.IpAddress) {
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull`

second attempt with the IP code removed was

`You cannot call a method on a null-valued expression.
At line:6 char:27
+         'Power State' = Switch ($VM.PowerState) {
+                                 ~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:6 char:27
+         'Power State' = Switch ($VM.PowerState) {
+                                 ~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 
You cannot call a method on a null-valued expression.
At line:6 char:27
+         'Power State' = Switch ($VM.PowerState) {
+                                 ~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull`

**Comment by tpcarman** [comment]
This issue should be resolved in 5a2992c. Please reopen if you continue to experience issues.
